### ISSUE TITLE
[Feature Request] JSON-RPC method for calling Move functions or execute Move code.

### ISSUE THREAD

**Issue by xli** [diem/diem#7551]
# 🚀 Feature Request

A JSON-RPC method that can call a Move functions defined in stdlib or modules that will be used by transaction script, returns the function call result.
It is also acceptable to define a specific set of functions for this method to call.

## Motivation

* We have some calculation logic used by transaction script, like calculate non-XDX currency travel rule limit threshold. To give a client accurate number of the threshold number, we should have a way to invoke same logic for the client built by different language.
* Explore options like https://web3js.readthedocs.io/en/v1.2.11/web3-eth-contract.html

## Pitch

TBD

**Describe alternatives you've considered**

**Are you willing to open a pull request?** (See [CONTRIBUTING](../../CONTRIBUTING.md))

## Additional context

Current travel rule limit threshold is defined for XDX only, for different currency, you need convert the amount into XDX currency amount first for comparing with the threshold. The calculation involves divide, thus may cause rounding difference if we tried to replicate same logic into different languages.

Client need know exact number of the threshold number for deciding whether the payment requires KYC data exchange with the other VASP.

**Comment by sblackshear** [comment]
CC @jolestar, who I think has mentioned this in the past.

**Comment by sblackshear** [comment]
@vgao1996, are there any blockers left for implementing this now that the `execute_function` refactoring is done?

**Comment by jolestar** [comment]
How about the RPC API designed?  The json RPC method can call any Move function? if the function change the state, should return an error or return the state and function result both? eth only support call read-only function.

**Comment by xli** [comment]
> How about the RPC API designed? The json RPC method can call any Move function? if the function change the state, should return an error or return the state and function result both? eth only support call read-only function.

support read-only function is fine.

I think the best will be given input, move function, get output, the move function can only operate on the input.

**Comment by vgao1996** [comment]
> are there any blockers left for implementing this now that the execute_function refactoring is done?

@sblackshear Sorry for the late reply. I don't think there are road blockers left.

I do have a few questions however:
1. Should this be metered? If so, how?
2. Overall this seems to be a very powerful feature. Is it desired to expose the Move layer directly to clients? How would this compare with the option of adding a new interface to the Diem adapter, which the clients can interact with, and that the Diem adapter will use a Move VM internally to get the result. Would the latter achieve better encapsulation?
3. This also doesn't really look that different from our existing scripts & function scripts -- maybe another option is to allow them to return values to clients?

**Comment by sblackshear** [comment]
> I think the best will be given input, move function, get output, the move function can only operate on the input.

"Input" includes both the function arguments and the current on-chain state, correct?

Also: I think it's quite reasonable to start with this, but in the long run we might want to support lifting the read-only restriction + returning the WriteSet from the function call as well. That could be a powerful dev tool, but isn't needed for the initial use-cases.

> Should this be metered? If so, how?

I think this is important for the long term (since in the presence of custom modules, an author can expose a non-terminating endpoint), but not as important in the short term for Diem because our exposed functions are very limited. I would suggest an approach where we set a fairly low `max_gas_units` on the server side (high enough to accommodate use-cases like reading on-chain constants, but not much more) and consider increases as we see the need.

> Overall this seems to be a very powerful feature. Is it desirable to expose the Move layer directly to clients? How would this compare with the option of adding a new interface to the Diem adapter, which the clients can interact with, and that the Diem adapter will use a Move VM internally to get the result? Would the latter achieve better encapsulation?
> This also doesn't really look that different from our existing scripts & function scripts -- maybe another option is to allow them to return values to clients?

I think directly exposing an API for calling Move functions is desirable. The way I think about this is:
- The Diem adapter layer takes transactions as input and produces writes to the ledger as output
- The new API we will expose takes functions as input and returns values as output (no writes or side effects of any kind)

Because the inputs and the roles of these layers are so different, it makes sense to expose them via separate APIs. I don't think encapsulation is a concern for this new layer because it is read-only; it is only a more convenient way of exposing information that clients already had access to.

**Comment by vgao1996** [comment]
The separation of roles makes sense to me! Just one question: how do we enforce that these queries are actually read-only? I can think of two ways:
1. We simply don't commit the write set.
2. Implement another verifier check disallowing `borrow_global_mut`, `move_to` and `move_from`.

**Comment by xli** [comment]
I never expect this function execution can change anything on-chain or local database data.
So I should state it in this way: readonly is required.

**Comment by sblackshear** [comment]
Yes, totally agree that function execution can't modify anything! What I was suggesting is that because of Move's execution model (which outputs a WriteSet instead of performing modifications on-the-fly), we could still choose to allow users to execute Move functions that perform writes and return the WriteSet to the user. Such an execution would be read-only from the perspective of json-rpc, but allow a user to see the writes that a function would perform.

**Comment by jolestar** [comment]
We implement an execute_readonly_function in starcoin base MoveVM

https://github.com/starcoinorg/starcoin/blob/9ce8fc03d96461ab6b8fbcdf58b1e533b2501917/vm/vm-runtime/src/starcoin_vm.rs#L830

if the function change the state will report an error.  I think this can be a Move VM feature.
